src/control/cmd/daos_agent/procmon.go

```diff
@@ -0,0 +1,209 @@
+//
+// (C) Copyright 2020 Intel Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"syscall"
+	"time"
+
+	"github.com/daos-stack/daos/src/control/drpc"
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+type procMonRequest struct {
+	// Pid of the process that the request is for
+	pid int32
+	// Whether the message is an attach or disconnect. Uses drpc method identifiers.
+	action drpc.MgmtMethod
+}
+
+type procMonResponse struct {
+	// Pid the response is coming from
+	pid int32
+	// Error indicating why the process "died"
+	err error
+}
+
+type procInfo struct {
+	log       logging.Logger
+	pid       int32
+	ctx       context.Context
+	cancelCtx func()
+	response  chan *procMonResponse
+}
+
+// NewProcInfo creates a new procInfo struct for use by procMon. The context passed
+// in is the top level context which a child context and cancellation routine is generated.
+// The response channel is the passed in by procMon to provide a unified interface
+// in procMon to handle responses.
+func NewProcInfo(ctx context.Context, Log logging.Logger, Pid int32, Response chan *procMonResponse) *procInfo {
+	child, cancel := context.WithCancel(ctx)
+	return &procInfo{
+		log:       Log,
+		pid:       Pid,
+		ctx:       child,
```

Comment:
If we're not doing any setup in the NewProcInfo why have one? The only reason we kept NewProcInfo was because of the context creation from the child and obtaining the cancelation function.

Reply:
I don't feel strongly about keeping the constructor. Stashing the cancel function directly in the struct is fine too.

Reply:
Removed it since it was only used once.